### ISSUE TITLE
[BUG]: groupkfold issue: string as fold_groups causes automl to fail

### ISSUE THREAD

**Issue by Ali-Flt** [pycaret/pycaret#3072]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [X] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

Even after the latest PR #3057 and installing from the master branch, I tried using a string (column name in train_data dataframe) as fold_groups input for GroupKFold but after running automl it returns error.
The error message makes me think it could be because the length of all data (train set + test set) is being compared to the length of only the train set somewhere.

### Reproducible Example

```python
from pycaret.regression import *
s = setup(
          train_data,
          target=label,
          test_data=test_data,
          imputation_type=None,
          normalize=True,
          index=True,
          html=True,
          profile=False,
          verbose=True,
          session_id=(3964),
          feature_selection=False,
          feature_selection_method='sequential',
          n_features_to_select=20,
          polynomial_features=False,
          polynomial_degree=1,
          remove_multicollinearity=True,
          fold_shuffle=True,
          fold_strategy='groupkfold',
          fold_groups='groups',
          fold=10,
          remove_outliers=False,
          transformation=False,
          ignore_features=['groups'],
          pca=True,
          pca_method='kernel',
          pca_components=30
          )

groups=train_data['groups']
search_library='optuna'
optimize='MAE'
tuner_verbose=0
bests = compare_models(turbo=False, n_select=4, sort=optimize, groups=groups)
leader_board = get_leaderboard()
best_model = automl(optimize=optimize, use_holdout=True)
```

### Expected Behavior

automl returning the best peforming model.

### Actual Results

```python-traceback
---------------------------------------------------------------------------
NotFittedError                            Traceback (most recent call last)
File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/internal/pycaret_experiment/supervised_experiment.py:5329, in _SupervisedExperiment.automl(self, optimize, use_holdout, turbo, return_train_score)
   5328 try:
-> 5329     self.predict_model(model, verbose=False)  # type: ignore
   5330 except Exception:

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/regression/oop.py:2214, in RegressionExperiment.predict_model(self, estimator, data, drift_report, round, verbose)
   2162 """
   2163 This function predicts ``Label`` using a trained model. When ``data`` is
   2164 None, it predicts label on the holdout set.
   (...)
   2211 
   2212 """
-> 2214 return super().predict_model(
   2215     estimator=estimator,
   2216     data=data,
   2217     probability_threshold=None,
   2218     encoded_labels=False,
   2219     drift_report=drift_report,
   2220     round=round,
   2221     verbose=verbose,
   2222 )

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/internal/pycaret_experiment/supervised_experiment.py:4985, in _SupervisedExperiment.predict_model(self, estimator, data, probability_threshold, encoded_labels, raw_score, drift_report, round, verbose, ml_usecase, preprocess)
   4983     estimator = get_estimator_from_meta_estimator(estimator)
-> 4985 pred = np.nan_to_num(estimator.predict(X_test_))
   4987 try:

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/sklearn/linear_model/_base.py:386, in LinearModel.predict(self, X)
    373 """
    374 Predict using the linear model.
    375 
   (...)
    384     Returns predicted values.
    385 """
--> 386 return self._decision_function(X)

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/sklearn/linear_model/_base.py:367, in LinearModel._decision_function(self, X)
    366 def _decision_function(self, X):
--> 367     check_is_fitted(self)
    369     X = self._validate_data(X, accept_sparse=["csr", "csc", "coo"], reset=False)

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/sklearn/utils/validation.py:1345, in check_is_fitted(estimator, attributes, msg, all_or_any)
   1344 if not fitted:
-> 1345     raise NotFittedError(msg % {"name": type(estimator).__name__})

NotFittedError: This LinearRegression instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator.

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
Cell In [20], line 60
     53 # best_tuned = []
     54 # for best in bests:
     55 #     try:
     56 #         best_tuned.append(tune_model(best, search_library=search_library, search_algorithm=search_algorithm , optimize=optimize, tuner_verbose=tuner_verbose, groups=groups))
     57 #     except:
     58 #         best_tuned.append(best)
     59 leader_board = get_leaderboard()
---> 60 best_model = automl(optimize=optimize, use_holdout=True)
     61 predict_results.append(predict_model(best_model))
     62 best_models.append(best_model)

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/utils/generic.py:957, in check_if_global_is_not_none.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
    955     if globals_d[name] is None:
    956         raise ValueError(message)
--> 957 return func(*args, **kwargs)

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/regression/functional.py:2260, in automl(optimize, use_holdout, turbo, return_train_score)
   2204 @check_if_global_is_not_none(globals(), _CURRENT_EXPERIMENT_DECORATOR_DICT)
   2205 def automl(
   2206     optimize: str = "R2",
   (...)
   2209     return_train_score: bool = False,
   2210 ) -> Any:
   2212     """
   2213     This function returns the best model out of all trained models in
   2214     current session based on the ``optimize`` parameter. Metrics
   (...)
   2257 
   2258     """
-> 2260     return _CURRENT_EXPERIMENT.automl(
   2261         optimize=optimize,
   2262         use_holdout=use_holdout,
   2263         turbo=turbo,
   2264         return_train_score=return_train_score,
   2265     )

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/regression/oop.py:2545, in RegressionExperiment.automl(self, optimize, use_holdout, turbo, return_train_score)
   2489 def automl(
   2490     self,
   2491     optimize: str = "Accuracy",
   (...)
   2494     return_train_score: bool = False,
   2495 ) -> Any:
   2497     """
   2498     This function returns the best model out of all trained models in
   2499     current session based on the ``optimize`` parameter. Metrics
   (...)
   2542 
   2543     """
-> 2545     return super().automl(
   2546         optimize=optimize,
   2547         use_holdout=use_holdout,
   2548         turbo=turbo,
   2549         return_train_score=return_train_score,
   2550     )

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/internal/pycaret_experiment/supervised_experiment.py:5334, in _SupervisedExperiment.automl(self, optimize, use_holdout, turbo, return_train_score)
   5330 except Exception:
   5331     self.logger.warning(
   5332         f"Model {model} is not fitted, running create_model"
   5333     )
-> 5334     model, _ = self._create_model(  # type: ignore
   5335         estimator=model,
   5336         system=False,
   5337         verbose=False,
   5338         cross_validation=False,
   5339         predict=False,
   5340         groups=self.fold_groups_param,
   5341         return_train_score=return_train_score,
   5342     )
   5343     self.pull(pop=True)
   5344     self.predict_model(model, verbose=False)  # type: ignore

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/internal/pycaret_experiment/supervised_experiment.py:1415, in _SupervisedExperiment._create_model(self, estimator, fold, round, cross_validation, predict, fit_kwargs, groups, refit, probability_threshold, experiment_custom_tags, verbose, system, add_to_model_list, X_train_data, y_train_data, metrics, display, model_only, return_train_score, **kwargs)
   1412             data_X = self.X_train
   1413     data_y = self.y_train if y_train_data is None else y_train_data.copy()
-> 1415 groups = self._get_groups(groups, data=data_X)
   1417 if metrics is None:
   1418     metrics = self._all_metrics

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/internal/pycaret_experiment/tabular_experiment.py:107, in _TabularExperiment._get_groups(self, groups, data, fold_groups)
    105 data = data if data is not None else self.X_train
    106 fold_groups = fold_groups if fold_groups is not None else self.fold_groups_param
--> 107 return pycaret.utils.generic.get_groups(groups, data, fold_groups)

File /opt/miniconda3/envs/pycaret_env_dev/lib/python3.8/site-packages/pycaret/utils/generic.py:815, in get_groups(groups, X_train, default)
    813 else:
    814     if groups.shape[0] != X_train.shape[0]:
--> 815         raise ValueError(
    816             f"groups has length {groups.shape[0]} which doesn't match X_train length of {len(X_train)}."
    817         )
    819 return groups

ValueError: groups has length 182 which doesn't match X_train length of 128.
```

### Installed Versions

<details>
System:
    python: 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:18)  [GCC 10.3.0]
executable: /opt/miniconda3/envs/pycaret_env_dev/bin/python3.8
   machine: Linux-6.0.2-arch1-1-x86_64-with-glibc2.10

PyCaret required dependencies:
                 pip: 22.3.1
          setuptools: 65.5.1
             pycaret: 3.0.0rc4
             IPython: 8.6.0
          ipywidgets: 8.0.2
                tqdm: 4.64.1
               numpy: 1.22.4
              pandas: 1.4.4
              jinja2: 3.1.2
               scipy: 1.8.1
              joblib: 1.2.0
             sklearn: 1.1.3
                pyod: 1.0.6
            imblearn: 0.9.1
   category_encoders: 2.5.1.post0
            lightgbm: 3.3.3
               numba: 0.55.2
            requests: 2.28.1
          matplotlib: 3.6.2
          scikitplot: 0.3.7
         yellowbrick: 1.5
              plotly: 5.11.0
             kaleido: 0.2.1
         statsmodels: 0.13.5
              sktime: 0.13.2
               tbats: 1.1.1
            pmdarima: 1.8.5
              psutil: 5.9.3

PyCaret optional dependencies:
                shap: Not installed
           interpret: Not installed
                umap: Not installed
    pandas_profiling: Not installed
  explainerdashboard: Not installed
             autoviz: Not installed
           fairlearn: Not installed
             xgboost: Not installed
            catboost: Not installed
              kmodes: Not installed
             mlxtend: Not installed
       statsforecast: Not installed
        tune_sklearn: 0.4.4
                 ray: 2.0.1
            hyperopt: 0.2.7
              optuna: 3.0.3
               skopt: 0.9.0
              mlflow: 1.30.0
              gradio: 3.9
             fastapi: 0.86.0
             uvicorn: 0.19.0
              m2cgen: 0.10.0
           evidently: 0.1.59.dev3
                nltk: Not installed
            pyLDAvis: Not installed
              gensim: Not installed
               spacy: Not installed
           wordcloud: Not installed
            textblob: Not installed
               fugue: Not installed
           streamlit: Not installed
             prophet: Not installed
</details>

**Comment by moezali1** [comment]
@Ali-Flt Can you please provide reproducible Notebook along with the sample data?

**Comment by tvdboom** [comment]
I fixed this issue here #3070. Please, wait for the merge and try again
